### PR TITLE
DSDEGP-2013: arrays integrated tests

### PR DESCRIPTION
**Pull request by pshapiro4broad**
DSDEGP-2013: Add support for integrated tests for the arrays index

- fix bug uncovered by tests where `query-arrays` would fail due to incorrect encoding of the chipwell-barcode query term
- in addition to the usual tests, also test that a missing idat will cause delivery to fail

---

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: Ensure that you have added or updated tests
- [x] **Submitter**: If you're adding/switching libraries or otherwise changing the design, update the [design documentation](https://broadinstitute.atlassian.net/wiki/pages/viewpage.action?pageId=114531509).
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
* Review cycle:
  * Team may comment on PR at will
  * **Reviewer assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to reviewer** for further feedback
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Move the JIRA issue to QA once this checklist is completed

### CODE REVIEW

File: clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/ArraysTests.scala [hunk 1]
```diff
@@ -0,0 +1,1040 @@
...
+    val vcfSource = rootSource / vcfName
+    val vcfIndexSource = rootSource / vcfIndexName
+    val fingerprintingDetailMetricsSource = rootSource / fingerprintingDetailMetricsName
+    val fingerprintingSummaryMetricsSource = rootSource / fingerprintingSummaryMetricsName
+
+    val endBasename = if (changeBasename) randomId else barcode
+
+    val rootDestination = rootSource.parent / s"moved/$id/"
+    val vcfDestination = rootDestination / s"$endBasename${ArraysExtensions.VcfGzExtension}"
+    val craiDestination = rootDestination / s"$endBasename${ArraysExtensions.VcfGzTbiExtension}"
```
Comment (danxmoran) [inline clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/ArraysTests.scala]: s/crai/tbi/

File: clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/ArraysTests.scala [hunk 2]
```diff
@@ -0,0 +1,1040 @@
...
+    workspaceName: Option[String] = None
+  ): Future[Assertion] = {
+    val deleteNote = s"$randomId --- Deleted by the integration tests --- $randomId"
+
+    val id = randomId
+    val barcode = s"barcode$id"
+    val version = 3
+
+    val vcfContents = s"$id --- I am a vcf fated to die --- $id"
+    val vcfIndexContents = s"$id --- I am an index fated to die --- $id"
```
Comment (danxmoran) [inline clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/ArraysTests.scala:542]: We should assert on [all the files-to-be-deleted](https://github.com/broadinstitute/clio/blob/develop/clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/arrays/ArraysMetadata.scala#L48), not just the vcf.

File: clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/ArraysTests.scala [hunk 3]
```diff
@@ -0,0 +1,1040 @@
...
+
+    val vcfContents = s"$id --- I am a dummy vcf --- $id"
+    val vcfIndexContents = s"$id --- I am a dummy vcfIndex --- $id"
+    val fingerprintingDetailMetricsContents =
+      s"$id --- I am dummy fingerprintingDetail metrics --- $id"
+    val fingerprintingSummaryMetricsContents =
+      s"$id --- I am dummy fingerprintingSummary metrics --- $id"
+
+    val vcfName = s"$barcode${ArraysExtensions.VcfGzExtension}"
+    val vcfIndexName =
```
Comment (danxmoran) [inline clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/ArraysTests.scala:375]: We should check that the gtc is moved, too.

File: clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/ArraysTests.scala [hunk 4]
```diff
@@ -0,0 +1,1040 @@
...
+    force = true
+  )
+
+  it should "move files and record the workspace name when delivering" in {
+    val id = randomId
+    val barcode = s"barcode$id"
+    val version = 3
+
+    val vcfContents = s"$id --- I am a dummy vcf --- $id"
+    val vcfIndexContents = s"$id --- I am a dummy vcfIndex --- $id"
```
Comment (danxmoran) [inline clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/ArraysTests.scala:675]: These delivery tests should include the gtc, too.

File: clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/DockerIntegrationSpec.scala [hunk 5]
```diff
@@ -100,6 +100,10 @@ class CoreDockerUbamSpec extends DockerIntegrationSpec("Clio in Docker") with Ub
 class CoreDockerCramSpec extends DockerIntegrationSpec("Clio in Docker") with WgsCramTests
 class CoreDockerGvcfSpec extends DockerIntegrationSpec("Clio in Docker") with GvcfTests
 
+class CoreDockerArraysSpec
+    extends DockerIntegrationSpec("Clio in Docker")
+    with ArraysTests
+
```
Comment (tlangs) [inline clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/DockerIntegrationSpec.scala:106]: Why is this formatted differently from the lines above it?

**Comment by pshapiro4broad** [comment]
The integration tests were run here: https://gotc-jenkins.dsp-techops.broadinstitute.org/job/clio-integration/354/
The only failure was due to version skew since the docker image is different than the current dev deployment:
```
[info] - should report the server version of this test at /version *** FAILED ***
[info]   VersionInfo(1.9.0-g5dc8eae-SNAP) was not equal to VersionInfo(1.9.0-g71f1fdb-SNAP) (BasicTests.scala:19)
```

File: clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/ArraysTests.scala (hunk 1 above)
Comment (tlangs) [inline clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/ArraysTests.scala]: should be `vcfIndexDestination`

File: clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/ArraysTests.scala [hunk 6]
```diff
@@ -0,0 +1,1040 @@
...
+    val vcfIndexContents = s"$id --- I am an index fated to die --- $id"
+    val metrics1Contents = s"$id --- I am an immortal metrics file --- $id"
+    val metrics2Contents =
+      s"$id --- I am a second immortal metrics file --- $id"
+
+    val storageDir = rootTestStorageDir / s"arrays/$barcode/v$version/"
+    val vcfPath = storageDir / s"$randomId${ArraysExtensions.VcfGzExtension}"
+    val vcfIndexPath = storageDir / s"$randomId${ArraysExtensions.VcfGzTbiExtension}"
+    val metrics1Path = storageDir / s"detail-$id.metrics"
+    val metrics2Path = storageDir / s"summary-$id.metrics"
```
Comment (tlangs) [inline clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/ArraysTests.scala]: we should also probably assert that all files that are not supposed to be deleted are left alone.

**Review by tlangs — COMMENTED** [review]
We seem to be redefining a lot of constants in each test. Would it make more sense to define some top-level constants like chip type, version, etc and reuse those in each test?

File: clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/DockerIntegrationSpec.scala (hunk 5 above)
Comment (pshapiro4broad) [reply clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/DockerIntegrationSpec.scala:106]: only scalafmt knows

File: clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/ArraysTests.scala (hunk 6 above)
Comment (pshapiro4broad) [reply clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/ArraysTests.scala]: I think that's what
```
        Seq((metrics1Path, metrics1Contents), (metrics2Path, metrics2Contents)).foreach {
          case (path, contents) =>
            path should exist
            path.contentAsString should be(contents)
        }
```
below is doing

File: clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/DockerIntegrationSpec.scala (hunk 5 above)
Comment (tlangs) [reply clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/DockerIntegrationSpec.scala:106]: sigh

**Comment by pshapiro4broad** [comment]
> We seem to be redefining a lot of constants in each test. Would it make more sense to define some top-level constants like chip type, version, etc and reuse those in each test?

In most cases the values contain random keys and are generated for each test, in part to avoid database collisions. It would be better to have a factory method that creates the various bits used by each test, but the needs for each test vary slightly. I'll see what I can do.

**Comment by pshapiro4broad** [comment]
Am going to hold off on refactoring the test code; there is already a JIRA for this work, DSDEGP-1796, and it's probably a medium task on its own.

**Comment by pshapiro4broad** [comment]
I think I've addressed the PR comments @danxmoran @tlangs

**Review by tlangs — APPROVED** [review]

**Review by danxmoran — APPROVED** [review]

**Pull request merged**